### PR TITLE
Updated bshd

### PR DESCRIPTION
**Pull request by xintin**
Updated the PR to mimic PyTorch's `torch.nn.functional.scaled_dot_product_attention` with optional causal masking and improved numerical stability. Intended for comparison and debugging purposes.

### CODE REVIEW

File: tests/kernel/wave/attention/vanilla_attention_test.py [hunk 1]
```diff
@@ -295,6 +295,93 @@ def testAttentionCausal(
...
+            q.transpose(1, 2).contiguous(),
+            k.transpose(1, 2).contiguous(),
+            v.transpose(1, 2).contiguous(),
+            output,
+        )
+        assert_close(
+            output.transpose(1, 2),
+            torch_ref,
+            check_dtype=False,
+            atol=1e-2,  # setting it to 1e-2 to tolerate minor fp16 softmax instability.
```
Comment (raikonenfnu) [inline tests/kernel/wave/attention/vanilla_attention_test.py:380]: Can you double check this comment, the softmax should be computed in fp32 since it is post mfma

File: tests/kernel/wave/common/utils.py [hunk 2]
```diff
@@ -37,3 +40,40 @@ def param_bool(name, shortname=None, values=None):
...
+        key (Tensor): key tensor of shape [B, H, S_k, D].
+        value (Tensor): value tensor of shape [B, H, S_k, D].
+        is_causal (bool): If True, applies causal masking to the attention logits.
+
+    Returns:
+        Tensor: Output tensor of shape [B, H, S_q, D] after applying attention.
+    """
+    scale: float = query.shape[-1] ** -0.5
+    attn_logits: Tensor = torch.matmul(query, key.transpose(-2, -1)) * scale
+
```
Comment (raikonenfnu) [inline tests/kernel/wave/common/utils.py:64]: as discussed offline, let's try use f32 softmax, maybe it would be closer to our wave output :)

Comment (xintin) [reply tests/kernel/wave/common/utils.py:71]: I have added checks for the `dtype `and `update to fp32` accordingly.

File: iree/turbine/kernel/wave/templates/vanilla_attention.py [hunk 3]
```diff
@@ -192,3 +193,270 @@ def base_attention(
...
+            acc = tkw.mma(v_reg, imm_f16, new_acc)
+            return m_j, d_j, acc
+
+        # repeat represents the results of the loop
+        res_max, res_sum, res_mm = repeat
+        reciprocal_sum = tkw.reciprocal(res_sum)
+        res = res_mm * reciprocal_sum
+        tkw.write(res, c, mapping=mapping, elements_per_thread=STORE_ELEMS_PER_THREAD)
+
+    @tkw.wave(constraints)
```
Comment (xintin) [inline iree/turbine/kernel/wave/templates/vanilla_attention.py:401]: I will fix the redundant code once the tests pass.
